crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -71,101 +72,167 @@ impl CoinGecko {
+            .collect();
+
+        Ok(parsed_response)
+    }
+
+    /// Fetches the prices of the given tokens denominated in the given token.
+    async fn bulk_fetch_denominated_in_token(
+        &self,
+        mut tokens: Vec<&Token>,
+        denominator: Token,
+    ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        tokens.push(&denominator);
+        let prices_in_eth = self.bulk_fetch_denominated_in_eth(&tokens).await?;
+
+        // fetch price of token we want to denominate all other prices in
+        let denominator_price: Decimal = prices_in_eth
+            .get(&denominator)
+            .cloned()
+            .ok_or(PriceEstimationError::NoLiquidity)?
+            .try_into()
+            .context("failed to convert price to decimal")?;
+
+        let prices_in_denominator = tokens
+            .into_iter()
+            .map(|token| {
+                let result = Self::denominate_price(token, denominator_price, &prices_in_eth);
+                (*token, result)
+            })
+            .collect();
+
+        Ok(prices_in_denominator)
+    }
+
+    /// CoinGecko provides all prices denominated in ETH.
+    /// This function converts such a token price to a price denominated
+    /// in a token provided by the caller.
+    fn denominate_price(
+        token: &Token,
+        denominator_price_eth: Decimal,
+        prices: &HashMap<Token, f64>,
+    ) -> NativePriceEstimateResult {
+        let token_price_eth: Decimal = prices
+            .get(token)
+            .cloned()
+            .ok_or(PriceEstimationError::NoLiquidity)?
+            .try_into()
+            .context("failed to convert price to decimal")?;
+
+        Ok(token_price_eth
+            .checked_div(denominator_price_eth)
+            .context("division by zero")?
+            .to_f64()
+            .context("failed to convert price back to f64")?)
+    }
+}
+
+#[async_trait]
+impl NativePriceBatchFetcher for CoinGecko {
+    async fn fetch_native_prices(
+        &self,
+        requested_tokens: &HashSet<H160>,
+    ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        let mut tokens = requested_tokens.iter().collect::<Vec<_>>();
+        match self.quote_token {
+            QuoteToken::Eth => {
+                let prices = self.bulk_fetch_denominated_in_eth(&tokens).await?;
+                Ok(tokens
+                    .into_iter()
+                    .map(|token| {
+                        let result = prices
+                            .get(token)
+                            .cloned()
+                            .ok_or(PriceEstimationError::NoLiquidity);
+                        (*token, result)
+                    })
+                    .collect())
+            }
+            QuoteToken::Other(native_price_token) => {
+                if !requested_tokens.contains(&native_price_token) {
+                    tokens.push(&native_price_token);
```

Comment:
careful here with the `token` size

Reply:
Yes, I'd say the responsibility lies on the config to provide 19 as the batch size.